src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,149 @@
+    private Node<T> tail;
+    private int size;
+
+    private static class Node<T> {
+        private Node<T> prev;
+        private T value;
+        private Node<T> next;
+
+        public Node(Node<T> prev, T value, Node<T> next) {
+            this.prev = prev;
+            this.value = value;
+            this.next = next;
+        }
+    }
+
     @Override
     public void add(T value) {
+        Node<T> tailNew = new Node<>(null, value, null);
+        if (isEmpty()) {
+            tail = tailNew;
+            head = tail;
+        } else if (size >= 1) {
+            tailNew.prev = tail;
+            tail.next = tailNew;
+            tail = tailNew;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        ifIndexOutOfBounds(index, size);
+        if (size <= 1 || index == size) {
+            add(value);
+            return;
+        } else if (index == 0) {
+            head.prev = new Node<>(null, value, head);
+            head = head.prev;
+        } else {
+            Node<T> current = getNodeByIndex(index);
+            Node<T> nodeNew = new Node<>(current.prev, value, current);
+            nodeNew.prev.next = nodeNew;
+            current.prev = nodeNew;
+        }
+        size++;
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        Node<T> current = head;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+        return current;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T entry : list) {
+            add(entry);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        ifIndexOutOfBounds(index, size - 1);
+        if (index == 0) {
+            return head.value;
+        }
+        Node<T> current = getNodeByIndex(index);
+        return current.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        ifIndexOutOfBounds(index, size - 1);
+        T valueToReturn = head.value;
+        if (index == 0) {
+            head.value = value;
+            return valueToReturn;
+        }
+        Node<T> current = getNodeByIndex(index);
+        valueToReturn = current.value;
+        current.value = value;
+        return valueToReturn;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        ifIndexOutOfBounds(index, size - 1);
+        Node<T> current = getNodeByIndex(index);
+        T valueToReturn = current.value;
+        unlink(current);
+        size--;
+        return valueToReturn;
     }
 
     @Override
     public boolean remove(T object) {
+        if (isEmpty()) {
+            return false;
+        }
+        Node<T> current = head;
+        for (int i = 0; i < size; i++) {
+            if (current.value == object || (current.value != null
+                    && current.value.equals(object))) {
+                unlink(current);
+                size--;
```

Comment:
why not to move `size--` inside `unlink(..)` method? 
The unlink method is responsible for deleting nodes, so the size should always be decreased in such cases.